ADR036: Arbitrary signature (#7727)
cosmos/cosmos-sdk: docs/architecture/adr-036-arbitrary-signature.md

# ADR 036: Arbitrary Message Signature Specification

## Changelog

- 28/10/2020 - Initial draft

## Authors
- Antoine Herzog (@antoineherzog)
- Zaki Manian (@zmanian)
- Aleksandr Bezobchuk (alexanderbez) [1]

> - Frojdi Dymylja (@fdymylja)

Reviewer: Can you please add me to the Authors list if you accept my new format and justification?

`- Ethan Frey (@ethanfrey)`

## Status

Draft

## Abstract

> Currently, in the SDK, there is no convention to sign arbitrary message like on Ethereum. We propose with this specification a way to sign arbitrary message for Cosmos SDK chains.

Reviewer 2: ```suggestion
Currently, in the SDK, there is no convention to sign arbitrary message like on Ethereum. We propose with this specification, for Cosmos SDK ecosystem, a way to sign and validate off-chain arbitrary messages.
```

## Context

Having the ability to sign messages off-chain has proven to be a fundamental aspect of nearly any blockchain. The notion of signing messages off-chain has many added benefits such as saving on computational costs and reducing transaction throughput and overhead. Within the context of the Cosmos, some of the major applications of signing such data includes, but is not limited to, providing a cryptographic secure and verifiable means of proving validator identity and possibly associating it with some other framework or organization. In addition, having the ability to sign Cosmos messages with a Ledger or similar HSM device.

## Decision

The aim is being able to sign arbitrary messages, even using Ledger or similar HSM devices.

> As a result signed messages should look roughly like Cosmos SDK messages. chain-id, account_number and sequence can all be assigned invalid values.

Reviewer 2: ```suggestion
As a result signed messages should look roughly like Cosmos SDK messages but **must not** be a valid on-chain transaction. `chain-id`, `account_number` and `sequence` can all be assigned invalid values.
```

The CLI should set those to default values.

Cosmos SDK 0.40 also introduces a concept of “auth_info” this can specify SIGN_MODES.

> A spec should include an auth info that supports SIGN_MODE_DIRECT and SIGN_MODE_LEGACY_AMINO.

Reviewer 2: ```suggestion
A spec should include an `auth_info` that supports SIGN_MODE_DIRECT and SIGN_MODE_LEGACY_AMINO.
```

- the memo should be empty
- nonce, sequence number should be equal to 0
- chain-id should be equal to “signature”
- fee gas should be equal to 0
- fee amount should be an empty array

> - Inside the message with the type MsgSignText, we put inside a  *bytes* message and the address of the signer.

Reviewer 2: let's use MUST and MUST NOT in the spec above where possible.

Also, how about rewordgin the lines 33-42 and putting `auth_info` into the enumerated list, to have clear bullet points of requirements.

Author: I will address the changes

Proto definition:
```proto
// MsgSignedMessage defines 

> message MsgSignedMessage {

Reviewer 3: I believe this should be `MsgSignText` based on the previous text.

Reviewer 2: I  would prefer `MsgSignData`. Msg is overloaded here (3 times in 2 "words")

    // Signer is the sdk.AccAddress of the message signer
    bytes Signer = 1 [(gogoproto.jsontag) = "signer", (gogoproto.casttype) = "github.com/cosmos/cosmos-sdk/types.AccAddress"];
    // Message represents the raw bytes of the content that is signed (text, json, etc)

>     bytes Message = 2 [(gogoproto.jsontag) = "message"];

Reviewer 3: How will this be displayed on Ledgers and similar HSMs?

Reviewer 4: I think at this we don't have an answer to that. Would you be OK with unblocking this on the promise that we'll continue doing work once this is approved?

Reviewer 2: The we should add a further discussion secton in the ADR and create a tracking issue on Github

Reviewer 4: > Then we should add a further discussion section in the ADR and create a tracking issue on Github

How could this reply of yours possibly answer my question?

Reviewer 5: Open an issue to facilitate further discussions. Let's get this merged. I would say the implementation is blocked until the issue on ledger and hsm is undersood and a solution has been agreed on. We can see if zondax can join a sdk call in the coming weeks.

Reviewer 4: OK, thanks @marbar3778 and @robert-zaremba

> }

Reviewer: I would propose the following, based on both the OAuth2 [implicit](https://tools.ietf.org/html/rfc6749#section-4.2.1) and [code grant flows](https://tools.ietf.org/html/rfc6749#section-4.1.1).

```proto
message MsgSignText {
    // we can add other request_types later with different semantics - this is for the obvious "prove your identity" / access control flow
    string request_type = 1;  // MUST be "bauth"
   
    // which client we are identifying with, could be `registry.atlas.io` or `2du6Sje7` but is unique to the site you authenticate with 
    string client_id = 2;     

    // (OPTIONAL) app-dependent if we want to give a partial grant - eg. for a forum, we could auth it to upvote but not make posts
    string scope = 3;

   // (OPTIONAL) a nonce or session token that can prevent replay attacks
   string state = 4;

    // ????? I don't think we need this, but it is in the Oauth2 flow - I belive this is different in the proposed flow.
    // but maybe is useful, I would like an end-to-end client workflow explained.
    string redirect_url = 5;
}
```

Note that the `Signer` would only be needed for actual processing in a module, via `GetSigners()`, but since we don't ever intend this to go to the blockchain, I would remove it as unnecessary data. The pubkey stored in the signature is the information we want to prove. 

This would allow a website to produce an unsigned transaction *without knowing your pubkey* and then you can sign it with any pubkey you control to identify yourself.

> }

Reviewer 2: How about adding the a `TTL` (time to live) field? Or do we want to have a minimum structure of `bytes Message` that all messages will comply?

Author: The general idea was this. This was studied to serve as a general purpose message that can basically fit in anything. The reasons being: a cosmos message has around it all the tooling required for signing and verification, and also arbitrary data can fit in itself every use case, for instance `Message` could be any type of structured data, a text, a file.

Reviewer 2: I can't imagine a use-case where we don't want TTL. But fair enough.
How about adding a discussion section to enumerate fields we think are necessary?

Author: A certificate which itself contains a TTL, plus there are different ways in which non replayability is achieved in systems, time isn't the only one. There are also nonce's systems etc.

Reviewer 2: OK, fair enough for now.

```
Signed MsgSignedMessage json example:
```json
{
  "type": "cosmos-sdk/StdTx",
  "value": {
    "msg": [
      {
        "type": "sign/MsgSignedMessage",
        "value": {
          "signer": "cosmos1hftz5ugqmpg9243xeegsqqav62f8hnywsjr4xr",
          "message": "cmFuZG9t"
        }
      }
    ],
    "fee": {
      "amount": [],
      "gas": "0"
    },
    "signatures": [
      {
        "pub_key": {
          "type": "tendermint/PubKeySecp256k1",
          "value": "AqnDSiRoFmTPfq97xxEb2VkQ/Hm28cPsqsZm9jEVsYK9"
        },
        "signature": "8y8i34qJakkjse9pOD2De+dnlc4KvFgh0wQpes4eydN66D9kv7cmCEouRrkka9tlW9cAkIL52ErB+6ye7X5aEg=="
      }
    ],
    "memo": ""
  }
}
```

## Consequences

> There is a specification on how messages, that are not meant to be broadcast to an active 

Reviewer 6: This seems like the sentence isn't finished (at least a work is missing)?

Author: addressed, thank you for pointing this out

## Context

Having the ability to sign messages off-chain has proven to be a fundamental aspect of nearly any blockchain. The notion of signing messages off-chain has many added benefits such as saving on computational costs and reducing transaction throughput and overhead. Within the context of the Cosmos, some of the major applications of signing such data includes, but is not limited to, providing a cryptographic secure and verifiable means of proving validator identity and possibly associating it with some other framework or organization. In addition, having the ability to sign Cosmos messages with a Ledger or similar HSM device.

## Decision

The aim is being able to sign arbitrary messages, even using Ledger or similar HSM devices.

As a result signed messages should look roughly like Cosmos SDK messages but **must not** be a valid on-chain transaction. `chain-id`, `account_number` and `sequence` can all be assigned invalid values.

Cosmos SDK 0.40 also introduces a concept of “auth_info” this can specify SIGN_MODES.

A spec should include an `auth_info` that supports SIGN_MODE_DIRECT and SIGN_MODE_LEGACY_AMINO.

- the memo must be empty

> - nonce, sequence number must be equal to 0

Reviewer 2: ```suggestion
- nonce, sequence number must be equal to 0 (this make the transaction invalid on-chain)
```

        "value": {
          "signer": "cosmos1hftz5ugqmpg9243xeegsqqav62f8hnywsjr4xr",
          "data": "cmFuZG9t"
        }
      }
    ],
    "fee": {
      "amount": [],
      "gas": "0"
    },
    "signatures": [
      {
        "pub_key": {
          "type": "tendermint/PubKeySecp256k1",
          "value": "AqnDSiRoFmTPfq97xxEb2VkQ/Hm28cPsqsZm9jEVsYK9"
        },
        "signature": "8y8i34qJakkjse9pOD2De+dnlc4KvFgh0wQpes4eydN66D9kv7cmCEouRrkka9tlW9cAkIL52ErB+6ye7X5aEg=="
      }
    ],
    "memo": ""
  }
}
```

## Consequences

There is a specification on how messages, that are not meant to be broadcast to a live chain, should be formed. 

### Backwards Compatibility

Backwards compatibility is maintained as this is a new message spec definition.

### Positive

- A common format that can be used by multiple applications to sign and verify off-chain messages.
- The specification is primitive which means it can cover every use case without limiting what is possible to fit inside it.
- It gives room for other off-chain messages specifications that aim to target more specific and common use cases such as off-chain-based authN/authZ layers [2].

### Negative

> 

Reviewer 2: ```suggestion

+ Current proposal requires a fixed relation between account address and a public key.
+ Doesn't work with multi sig accounts.
```

Reviewer 2: Please feel free to add more negative consequences - we were discussing it in the thread. Leaving it empty is not good.

> 

Reviewer 4: ```suggestion
## Further discussion
```

        "value": {
          "signer": "cosmos1hftz5ugqmpg9243xeegsqqav62f8hnywsjr4xr",
          "data": "cmFuZG9t"
        }
      }
    ],
    "fee": {
      "amount": [],
      "gas": "0"
    },
    "signatures": [
      {
        "pub_key": {
          "type": "tendermint/PubKeySecp256k1",
          "value": "AqnDSiRoFmTPfq97xxEb2VkQ/Hm28cPsqsZm9jEVsYK9"
        },
        "signature": "8y8i34qJakkjse9pOD2De+dnlc4KvFgh0wQpes4eydN66D9kv7cmCEouRrkka9tlW9cAkIL52ErB+6ye7X5aEg=="
      }
    ],
    "memo": ""
  }
}
```

## Consequences

There is a specification on how messages, that are not meant to be broadcast to a live chain, should be formed. 

### Backwards Compatibility

Backwards compatibility is maintained as this is a new message spec definition.

### Positive

- A common format that can be used by multiple applications to sign and verify off-chain messages.
- The specification is primitive which means it can cover every use case without limiting what is possible to fit inside it.
- It gives room for other off-chain messages specifications that aim to target more specific and common use cases such as off-chain-based authN/authZ layers [2].

### Negative

> + Current proposal requires a fixed relationship between an account address and a public key.

Reviewer 4: ```suggestion
- Current proposal requires a fixed relationship between an account address and a public key.
```

> + Doesn't work with multi sig accounts.

Reviewer 4: ```suggestion
- Doesn't work with multi sig accounts.

```

> ## Further discussion

Reviewer 4: ```suggestion

## Further discussion
```

          "data": "cmFuZG9t"
        }
      }
    ],
    "fee": {
      "amount": [],
      "gas": "0"
    },
    "signatures": [
      {
        "pub_key": {
          "type": "tendermint/PubKeySecp256k1",
          "value": "AqnDSiRoFmTPfq97xxEb2VkQ/Hm28cPsqsZm9jEVsYK9"
        },
        "signature": "8y8i34qJakkjse9pOD2De+dnlc4KvFgh0wQpes4eydN66D9kv7cmCEouRrkka9tlW9cAkIL52ErB+6ye7X5aEg=="
      }
    ],
    "memo": ""
  }
}
```

## Consequences

There is a specification on how messages, that are not meant to be broadcast to a live chain, should be formed. 

### Backwards Compatibility

Backwards compatibility is maintained as this is a new message spec definition.

### Positive

- A common format that can be used by multiple applications to sign and verify off-chain messages.
- The specification is primitive which means it can cover every use case without limiting what is possible to fit inside it.
- It gives room for other off-chain messages specifications that aim to target more specific and common use cases such as off-chain-based authN/authZ layers [2].

### Negative

- Current proposal requires a fixed relationship between an account address and a public key.

> - Doesn't work with multi sig accounts.

Reviewer 4: ```suggestion
- Doesn't work with multisig accounts.
```

# ADR 036: Arbitrary Message Signature

## Changelog

- 28/10/2020 - Initial draft

> ## Authors

Reviewer 3: This should reference some original work you can find [here](https://github.com/cosmos/ics/pull/33).

Author: addressed

- Antoine Herzog (@antoineherzog)
- Zaki Manian (@zmanian)
- Frojdi Dymylja (@fdymylja)

## Status

> Draft

Reviewer 3: ```suggestion

Draft
```

Author: addressed

## Abstract

Currently, in the SDK, there is no convention to sign arbitrary message like on Ethereum. We propose with this specification a way to sign arbitrary message for Cosmos SDK chains.

> ## Context

Reviewer 3: ```suggestion

## Context
```

Author: addressed

Having the ability to sign messages off-chain has proven to be a fundamental aspect of nearly any blockchain. The notion of signing messages off-chain has many added benefits such as saving on computational costs and reducing transaction throughput and overhead. Within the context of the Cosmos, some of the major applications of signing such data includes, but is not limited to, providing a cryptographic secure and verifiable means of proving validator identity and possibly associating it with some other framework or organization. In addition, having the ability to sign Cosmos messages with a Ledger or similar HSM device.

## Decision

The aim is being able to sign arbitrary messages, even using Ledger or similar HSM devices.

> As a result signed messages should look roughly like Cosmos SDK messages. chain-id, account_number and sequence can all be assigned invalid values.

Reviewer: :+1:

The CLI should set those to default values.

Cosmos SDK 0.40 also introduces a concept of “auth_info” this can specify SIGN_MODES.

A spec should include an auth info that supports SIGN_MODE_DIRECT and SIGN_MODE_LEGACY_AMINO.

- the memo should be empty
- nonce, sequence number should be equal to 0
- chain-id should be equal to “signature”
- fee gas should be equal to 20000 (0 is not applicable)
- fee amount should be an empty array
- Inside the message with the type MsgSignText, we put inside a  *bytes* message and the address of the signer.

## Consequences

There is a new optional module that offers arbitrary signing and verification capabilities.

### Backwards Compatibility

Backwards compatibility is maintained as this is a new module.

### Positive

- Ability to sign arbitrary messages

### Negative

- Probably this feature should be provided via the keystore 

> - Developing it as a module means we need to rely on using sign transactions in offline mode, and implementing it as default for the command means that we need to do a flag override (offline, account number, sequence, chain id)

Reviewer: Huh? This is a module? I thought this was just a format specification.

Can't you just define a format as above, and add a protobuf spec fo MsgSendText (and legacy amino compatibility for it)? The chain never needs to process this, as existing decorators would be sure to reject it.

Reviewer 7: Agreed, it could go in x/auth.
